staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go

```diff
@@ -0,0 +1,550 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package immutable
+
+import (
+	"fmt"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	xListTypeMap    = "map"
+	xListTypeAtomic = "atomic"
+	xListTypeSet    = "set"
+	xMapType        = "granular"
+
+	genericTypeInteger      = "integer"
+	genericTypeNumber       = "number"
+	genericTypeString       = "string"
+	genericTypeBool         = "boolean"
+	genericTypeObject       = "object"
+	genericTypeArray        = "array"
+	xImmutabilityImmutable  = "immutable"
+	xImmutabilityAddOnly    = "addOnly"
+	xImmutabilityRemoveOnly = "removeOnly"
+)
+
+// Immutable validates immutable values are respected or error otherwise
+func Immutable(old interface{}, new interface{}, s *structuralschema.Structural, pth *field.Path) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	switch old := old.(type) {
+	case map[string]interface{}:
+		newObj, ok := new.(map[string]interface{})
+		if !ok {
+			// e.g the CRD definition has changed
+			// TODO: what's the expectation here?
```

Comment:
It is a mutation of `oldObj`. So if  `s.XMutability` is set, we have to reject it.

For anything further inside we don't have to check anything.

Note, that this case could happen for raw JSON with `x-kubernetes-preserve-unknown-fields:true`.